solr/core/src/java/org/apache/solr/cloud/api/collections/CategoryRoutedAlias.java

```diff
@@ -119,13 +119,19 @@ public void validateRouteValue(AddUpdateCommand cmd) throws SolrException {
     }
 
     String dataValue = String.valueOf(cmd.getSolrInputDocument().getFieldValue(getRouteField()));
+    String mustMatch = aliasMetadata.get(ROUTER_MUST_MATCH);
     String candidateCollectionName = buildCollectionNameFromValue(dataValue);
     List<String> cols = getCollectionList(this.parsedAliases);
 
     if (cols.contains(candidateCollectionName)) {
       return;
     }
 
+    if (mustMatch != null && !candidateCollectionName.matches(mustMatch)) {
```

Comment:
We should use a pre-compiled expression. When the CRA is created this expression should be precompiled and stored as a field.